numba/core/types/functions.py

```diff
@@ -202,21 +356,63 @@ def get_impl_key(self, sig):
 
     def get_call_type(self, context, args, kws):
         template = self.template(context)
-        e = None
+        literal_e = None
+        nonliteral_e = None
+        _DEBUG = False
+
+
         # Try with Literal
         try:
             out = template.apply(args, kws)
-        except Exception:
+        except Exception as exc:
+            literal_e = exc
             out = None
-        # If that doesn't work, remove literals
-        if out is None:
-            args = [unliteral(a) for a in args]
-            kws = {k: unliteral(v) for k, v in kws.items()}
-            out = template.apply(args, kws)
-        if out is None and e is not None:
-            raise e
+
+        # if the unliteral_args and unliteral_kws are the same as the literal
+        # ones, set up to not bother retrying
+        unliteral_args = tuple([unliteral(a) for a in args])
+        unliteral_kws = {k: unliteral(v) for k, v in kws.items()}
+        skip = unliteral_args == args and kws == unliteral_kws
+
+        # If the above template application failed and the non-literal args are
+        # different to the literal ones, try again with literals rewritten as
+        # non-literals
+        if not skip and out is None:
+            try:
+                out = template.apply(unliteral_args, unliteral_kws)
+            except Exception as exc:
```

Comment:
need to let `errors.ForceLiteralArg` propagate up

Reply:
done